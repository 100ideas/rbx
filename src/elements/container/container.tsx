import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";

import { forwardRefAs, Generic, HelpersProps } from "../../base";
import { Variables } from "../../base/helpers/variables";

export type ContainerModifierProps = Partial<{
  breakpoint: Variables["Breakpoints"];
  fluid: boolean;
}>;

export type ContainerProps = HelpersProps & ContainerModifierProps;

const propTypes = {
  breakpoint: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  fluid: PropTypes.bool,
};

export const Container = Object.assign(
  forwardRefAs<ContainerProps, "div">(
    ({ className, fluid, breakpoint, ...rest }, ref) => (
      <Generic
        className={classNames(
          "container",
          {
            [`is-${breakpoint}`]: breakpoint,
            "is-fluid": fluid,
          },
          className,
        )}
        ref={ref}
        {...rest}
      />
    ),
    { as: "div" },
  ),
  { propTypes },
);
