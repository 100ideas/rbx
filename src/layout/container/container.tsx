import { cx } from "emotion";
import React from "react";

import { forwardRefAs } from "@/exotic";
import { ModifierProps, transformModifiers } from "@/modifiers";
import { Breakpoints } from "@/modifiers/responsive";

export type ContainerModifierProps = Partial<{
  breakpoint: Breakpoints;
  className: string;
  fluid: boolean;
}>;

export type ContainerProps = ModifierProps & ContainerModifierProps;

export const Container = forwardRefAs<ContainerProps, "div">((props, ref) => {
  const { as, fluid, breakpoint, ...rest } = transformModifiers(props);
  rest.className = cx("container", rest.className, {
    "is-fluid": fluid,
    [`is-${breakpoint}`]: breakpoint,
  });
  return React.createElement(as!, { ref, ...rest });
}, "div");

Container.defaultProps = Object.assign(
  { fluid: false },
  Container.defaultProps,
);
