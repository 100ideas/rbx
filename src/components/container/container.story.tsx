import React from "react";

import { storiesOf } from "@storybook/react";

import { Container } from "@/components/container";
import { Heading } from "@/components/heading";
import { Section } from "@/components/section";

storiesOf("Container", module).add("Default", () => (
  <div>
    <Section>
      <Container>
        <p className="bd-notification is-success">
          <Heading<"p"> size={5} as="p">
            Default
          </Heading>
          <Heading<"p"> subtitle as="p">
            Container
          </Heading>
        </p>
      </Container>
    </Section>
    <Section>
      <Container fluid>
        <p className="bd-notification is-info">
          <Heading<"p"> size={5} as="p">
            Fluid
          </Heading>
          <Heading<"p"> subtitle as="p">
            Container
          </Heading>
        </p>
      </Container>
    </Section>
    <Section>
      <Container breakpoint="widescreen">
        <p className="bd-notification is-warning">
          <Heading<"p"> size={5} as="p">
            Breakpoint Widescreen
          </Heading>
          <Heading<"p"> subtitle as="p">
            Container
          </Heading>
        </p>
      </Container>
    </Section>
    <Section>
      <Container breakpoint="fullhd">
        <p className="bd-notification is-danger">
          <Heading<"p"> size={5} as="p">
            Breakpoint Fullhd
          </Heading>
          <Heading<"p"> subtitle as="p">
            Container
          </Heading>
        </p>
      </Container>
    </Section>
  </div>
));
