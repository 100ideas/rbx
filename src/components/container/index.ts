import "./container.sass";

export { Container } from "./container";
