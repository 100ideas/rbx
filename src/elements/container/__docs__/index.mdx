---
name: Container
menu: Elements
route: /elements/container
---

import { Playground } from "docz";

import { Frame, SimplePropsTable, mapEnumerable } from "src/__docs__";
import { DEFAULTS } from "src/base/helpers/variables";
import { Notification } from "src/elements";
import { Container } from "../container";

# Container

## Container

The container centers its content horizontally.

<Playground>
  <Frame updateHeightDelay={0}>
    <Container>
      <Notification>
        This container is <strong>centered</strong> on desktop.
      </Notification>
    </Container>
  </Frame>
</Playground>

### Fluid

Use the `fluid` prop for the container to fit horizontally with padding.

<Playground>
  <Container fluid>
    <Notification>
      This container is <strong>fluid</strong>: it will have a 32px gap on
      either side, on any viewport size.
    </Notification>
  </Container>
</Playground>

### Breakpoint

<Playground>
  <Frame>
    {DEFAULTS.breakpoints.map(breakpoint => (
      <Container breakpoint={breakpoint} key={breakpoint}>
        <Notification>
          This container is <strong>fullwidth</strong> <em>until</em> the
          <code>{breakpoint}</code> breakpoint.
        </Notification>
      </Container>
    ))}
  </Frame>
</Playground>

## Properties

### Container

<SimplePropsTable
  props={{
    as: {
      defaultValue: "div",
      description: "the default element to render as",
      typeName: "React Type",
      typeTip: 'Any JSX Element (e.g. "div" or "span"), or a React Component'
    },
    breakpoint: {
      description:
        "the responsive size at which this container is no longer fullwidth",
      typeName: "string",
      typeTip: mapEnumerable(DEFAULTS.breakpoints)
    },
    fluid: {
      description:
        "whether the container expands to fit horizontally with padding",
      typeName: "boolean"
    }
  }}
/>

> The `as` prop enables components to wrap any other component.
> Therefore, they will also accept all props for the `as` component.

> These component also accept all props supported by the `Generic` component.
