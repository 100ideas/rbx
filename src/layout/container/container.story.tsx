import React from "react";

import { storiesOf } from "@storybook/react";

import { Title } from "@/elements";
import { Container } from "@/layout/container";
import { Section } from "@/layout/section";

storiesOf("Layout/Container", module).add("Default", () => (
  <div>
    <Section>
      <Container>
        <p className="bd-notification is-success">
          <Title<"p"> size={5} as="p">
            Default
          </Title>
          <Title<"p"> subtitle as="p">
            Container
          </Title>
        </p>
      </Container>
    </Section>
    <Section>
      <Container fluid>
        <p className="bd-notification is-info">
          <Title<"p"> size={5} as="p">
            Fluid
          </Title>
          <Title<"p"> subtitle as="p">
            Container
          </Title>
        </p>
      </Container>
    </Section>
    <Section>
      <Container breakpoint="widescreen">
        <p className="bd-notification is-warning">
          <Title<"p"> size={5} as="p">
            Breakpoint Widescreen
          </Title>
          <Title<"p"> subtitle as="p">
            Container
          </Title>
        </p>
      </Container>
    </Section>
    <Section>
      <Container breakpoint="fullhd">
        <p className="bd-notification is-danger">
          <Title<"p"> size={5} as="p">
            Breakpoint Fullhd
          </Title>
          <Title<"p"> subtitle as="p">
            Container
          </Title>
        </p>
      </Container>
    </Section>
  </div>
));
