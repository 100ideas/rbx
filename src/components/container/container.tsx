import { cx } from "emotion";
import React from "react";

import { Element, renderAsExoticComponent } from "@/components/element";
import { ModifierProps } from "@/modifiers";
import { Breakpoints } from "@/modifiers/responsives";

export type ContainerModifierProps = Partial<{
  children: React.ReactNode;
  fluid: boolean;
  breakpoint: Breakpoints;
}>;

export type ContainerProps = ModifierProps & ContainerModifierProps;

export const Container = renderAsExoticComponent<ContainerProps, "div">(
  ({ children, fluid, breakpoint, className, ...props }, ref) => (
    <Element
      {...props}
      ref={ref}
      className={cx("container", className, {
        "is-fluid": fluid,
        [`is-${breakpoint}`]: breakpoint,
      })}
    >
      {children}
    </Element>
  ),
  "div",
);
Container.defaultProps = Object.assign(
  {
    children: null,
    fluid: false,
  },
  Container.defaultProps,
);
