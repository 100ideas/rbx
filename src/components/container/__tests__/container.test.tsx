import React from "react";
import renderer from "react-test-renderer";

import { Container } from "../container";

describe("Container component", () => {
  it("should have container classname", () => {
    const component = renderer.create(
      <Container>
        <p className="bd-notification is-success">
          <p>Default</p>
          <p>Container</p>
        </p>
      </Container>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should have helper classnames", () => {
    const component = renderer.create(
      <Container
        paddingless
        responsive={{
          desktop: {
            hide: {
              value: true,
            },
          },
          touch: {
            display: {
              value: "flex",
            },
          },
          widescreen: {
            display: {
              only: true,
              value: "block",
            },
          },
        }}
      >
        <p className="bd-notification is-success">
          <p>Default</p>
          <p>Container</p>
        </p>
      </Container>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
