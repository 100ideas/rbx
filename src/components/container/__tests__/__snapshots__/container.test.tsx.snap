// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container component should have container classname 1`] = `
<div
  className="container"
>
  <p
    className="bd-notification is-success"
  >
    <p>
      Default
    </p>
    <p>
      Container
    </p>
  </p>
</div>
`;

exports[`Container component should have helper classnames 1`] = `
<div
  className="container is-paddingless is-hidden-desktop is-flex-touch is-block-widescreen-only"
>
  <p
    className="bd-notification is-success"
  >
    <p>
      Default
    </p>
    <p>
      Container
    </p>
  </p>
</div>
`;
